### ISSUE TITLE
OIDC Userfield not mapped since version 3.9.2 

### ISSUE THREAD

**Issue by Aldises** [microsoft/o365-moodle#1481]
Hi, 

I'm using the OIDC plugin without the Office065 plugin. My ID Provider is Keycloack.

Since the new version with the same configuration all the new users successfully logged into Moodle but they have to fill information about the name, lastname and email.

Did something change in the User Field mapping for this new version ?

All the best

**Comment by weilai-irl** [comment]
Hi @Aldises ,

Thank for your reporting this.

I think I know the reason why this happen. The latest release contains a change which removed the forced user profile updates for first name, last name, email, and ID number fields form ID provider. This change was only tested for Azure AD, but not for other ID providers. For Azure AD, additional field mapping settings are available in local_o365 plugins.

There is plan to implement standard field mapping functions in auth_oidc, but it's simply not there yet.

The best temporary solution now would be to create a reverse patch from commit https://github.com/microsoft/o365-moodle/commit/120417b7c3234fc5f923425afd3c9cd37e84d4b1, and apply it to your site until the standard field mapping settings in auth_oidc is available. Please give it a try and let me know if it solves the issue.

Regards,
Lai

**Comment by Aldises** [comment]
Hi @weilai-irl,

Thanks for your quick response, I've tried to reverse the patch with the old value. I've modified only the specific file and haven't touch the others. 

`  $forcedconfig = [           
            'field_updatelocal_idnumber' => 'oncreate',             
            'field_lock_idnumber' => 'locked',              
            'field_updatelocal_lang' => 'oncreate',           
            'field_lock_lang' => 'locked',              
            'field_updatelocal_firstname' => 'onlogin',           
            'field_lock_firstname' => 'unlocked',               
            'field_updatelocal_lastname' => 'onlogin',              
            'field_lock_lastname' => 'unlocked',            
            'field_updatelocal_email' => 'onlogin',       
            'field_lock_email' => 'unlocked',             
        ];`

Unfortunately, I'm still facing the same issue, shall I revert the entire plugin to an older version ? 

Do you think the standard field mapping will be implemented soon ? 

Best regards

**Comment by CoMPaTech** [comment]
@weilai-irl @Aldises not sure if you ever got this working again, but just stumbled on this myself

Reversing the patch was something I already tried from the code, seeing that `local_o365` just sets some of those fields to `onlogin`. Diffing the releases (v3.9.0 -> v3.9.2) of the plugin doesn't reveal much else, if I look at the tokens (and unpack the jwt token) all the information is there, it's just like it isn't picked up by Moodle.

Debugged `get_userinfo` but that fortunately does feed back the appropriate data. For the record, if you loop in `live log` I see 'user with ID {somenumber} logged in` and the `mdl_user` table does hold said user id with type oidc and the username being the uuid of the user (which is the same uuid the token has). So just for some reason the (pre)filling of the Firstname/Surname/Email address fields doesn't happen?

Tinkering with `field_lock_{fieldname}` works properly, filling it with 'onlogin' or 'oncreate' however doesn't prefill them.

In the end while typing up this reply I couldn't let it go and quiet actually it was pretty simple once I discovered how `update_user_record_by_id` within `lib/moodlelib.php` processed the data - eventually just never filling up those fields. So this isn't about the formentioned patch (though didn't try reverting that afterwards), but more likely due to some fields that don't update properly and/or a change on the datamodel by Moodle.

```
            $firstname = $idtoken->claim('given_name');
            if (!empty($firstname)) {
-                $userinfo['givenName'] = $firstname;
+                $userinfo['firstname'] = $firstname;
            }

            $lastname = $idtoken->claim('family_name');
            if (!empty($lastname)) {
-                $userinfo['surname'] = $lastname;
+                $userinfo['lastname'] = $lastname;
            }

            $email = $idtoken->claim('email');
            if (!empty($email)) {
-                $userinfo['mail'] = $email;
+                $userinfo['email'] = $email;
```

**Comment by Aldises** [comment]
@CoMPaTech Thanks for digging !

I'd had the time to look into it in details. I'm gonna try your solution today and see how it goes !

**Comment by weilai-irl** [comment]
Hi @Aldises and @CoMPaTech ,

I confirm this is an error introduced in the last release. @CoMPaTech is spot on in finding the root of the issue. In the get_userinfo() function, a change was made to the return value so that it's ready to be processed by local_o365 plugin to process additional field mapping fields. This doesn't work for sites that are using only auth_oidc but not local_o365.

I have addressed the issue now and a fix will be included in the next release before the end of January.

Regards,
Lai

**Comment by Aldises** [comment]
Hi @weilai-irl 

Thanks for the fix ! 

I'm looking forward for the official Release.

All the best

Aldises

**Comment by weilai-irl** [comment]
Hi @Aldises,

The fix to the issue has been included in the latest release 3.9.3 and 3.10.0 of the plugin. Please check it out.

I'll mark the issue closed for now. Please feel free to reopen it if you have questions about the fix.

Regards,
Lai
